torch/csrc/autograd/profiler.cpp

```diff
@@ -8,24 +8,61 @@
 #include <string>
 #include <vector>
 
+#include <ATen/ThreadLocalState.h>
+
 namespace torch { namespace autograd { namespace profiler {
 
 namespace {
 
 CUDAStubs default_stubs;
 constexpr CUDAStubs* default_stubs_addr = &default_stubs;
-// constant initialization, so it is guaranteed to be initialized before
+// Constant initialization, so it is guaranteed to be initialized before
 // static initialization calls which may invoke registerCUDAMethods
 static CUDAStubs* cuda_stubs = default_stubs_addr;
 
-ProfilerState state = ProfilerState::Disabled;
+thread_local ProfilerConfig config(ProfilerState::Disabled, false);
 // Protects access all_event_lists_map.
 std::mutex all_event_lists_map_mutex;
 std::unordered_map<uint16_t, std::shared_ptr<RangeEventList>>
     all_event_lists_map;
 thread_local std::shared_ptr<RangeEventList> event_list;
 thread_local uint16_t thread_id;
 
+bool unused_ = []() {
+  at::ThreadLocalState::registerThreadLocalSetting(
+    at::ThreadLocalSetting::PROFILER,
+    []() {
+      auto v = at::SettingValue();
+      v.pair.first = (int32_t)config.state;
+      v.pair.second = (int32_t)config.report_input_shapes;
+      return v;
+    },
+    [](at::SettingValue v) {
+      // ThreadLocalState propagates profiler state across threads and
+      // automatically enables/disables profiling in async tasks.
+      // Note, that in case of a subtask profiling, we don't
+      // want to consolidate profiling results but let the original profiler
+      // keep running
+      if (v.pair.first != (int32_t)ProfilerState::Disabled) {
+        enableProfiler(
+            ProfilerConfig{(ProfilerState)v.pair.first, (bool)v.pair.second},
+            /* emit_start */ false);
+      } else if (config.state != ProfilerState::Disabled) {
```

Comment:
so doesn't this function get called in destructor of thread-local guard here: https://github.com/pytorch/pytorch/blob/a5bfcc53237fa470e59604840be63c4abdacb875/aten/src/ATen/ThreadLocalState.h#L51

I can still come up with a contrived situation where this logic won't work:

Imagine that a thread has profiler already enabled and encounters ThreadLocalGuard with another thread state that also has profiler
* in Guard constructor we call enableProfiler because the new state is 'enabled', and we store the previous state which is also 'true'
* in Guard destructor we again call enableProfile again because the new (i.e. old) state is 'enabled'

So if I'm not mistaken we just bumped the counter by two :(

Basically asymmetry here makes it really hard to get the right logic. You either need to store both profiler state and depth or just ban reentrant profilers all together (might be a good idea :) )

Reply:
I see, yes that's a very valid point.

This is indeed very tricky and I'll try to simplify this. Let's use ThreadLocalDebugInfo struct and add a new slot in there - profiler data. Then each instance of profiler will operate on its instance of profiler data (just like other stuff within ThreadLocalDebugInfo, this instance of profiler data is propagated across threads and is shared by main thread and async tasks) and not the global set of thread_local event lists.

Then, this might make this case easier, as there will be less dependency between profilers.